Code/Tools/AssetProcessor/native/tests/assetmanager/AssetProcessorManagerTest.cpp

```diff
@@ -4219,21 +4222,33 @@ TEST_F(LockedFileTest, DeleteFile_LockedProduct_DeletesWhenReleased)
     ASSERT_GT(m_data->m_productPaths.size(), 0);
     QFile product(productPath);
 
+    // Open the file and keep it open to lock it
+    // We'll start a thread later to unlock the file
+    // This will allow us to test how AP handles trying to delete a locked file
     ASSERT_TRUE(product.open(QIODevice::ReadOnly));
 
     // Check if we can delete the file now, if we can't, proceed with the test
     // If we can, it means the OS running this test doesn't lock open files so there's nothing to test
     if (!AZ::IO::SystemFile::Delete(productPath.toUtf8().constData()))
     {
-        AZStd::thread workerThread;
+        AZStd::atomic_bool threadStarted = false;
 
-        m_callback = [&product, &workerThread]() {
-            workerThread = AZStd::thread([&product]() {
+        // Start a thread and wait some time before unlocking the file
+        // This should cause the AP to fail to delete the file at least once and force a retry
+        auto workerThread = AZStd::thread(
+            [&product, &threadStarted]()
+            {
+                threadStarted = true;
                 AZStd::this_thread::sleep_for(AZStd::chrono::milliseconds(60));
                 product.close();
             });
-        };
 
+        // Wait until we know the workerThread has started up
+        while (!threadStarted)
+        {
+            AZStd::this_thread::yield();
+        }
+        
```

Comment:
This is not performing the intended effect of the UnitTest.

This thread can start and complete, before the product file is deleted.
It is not testing in a deterministic way that the AssetProcessorManager fails to delete a locked file and then later succeeds.

Technically to perform that actual, the `AssessDeletedFile` signal below must be invoked, followed by waiting until the AssetProcessor is Idle, then this thread should be created and joined, followed by waiting until the AssetProcessor is Idle again

Reply:
The AP won't idle until its deleted though, is the problem.

The AP will become not-idle, then will loop a fixed number of times trying to deal with the file.  Then after that happens, it will give up and go idle and 'fail'.

This test is trying to figure out if it 'eventually succeeds' despite the file being locked for at least one iteration.

Reply:
I have a snippet of logic that should achieve the intended test scenario

```
QMetaObject::invokeMethod(
    m_assetProcessorManager.get(), "AssessDeletedFile", Qt::QueuedConnection, Q_ARG(QString, QString(theFileString)));

EXPECT_TRUE(BlockUntilIdle(5000));

// Start a thread and wait some time before unlocking the file
// This should cause the AP to fail to delete the file at least once and force a retry
auto workerThread = AZStd::thread(
    [&product]()
    {
        AZStd::this_thread::sleep_for(AZStd::chrono::milliseconds(60));
        product.close();
    });
workerThread.join();

EXPECT_TRUE(BlockUntilIdle(5000));

EXPECT_FALSE(QFile::exists(productPath));
EXPECT_EQ(m_data->m_deletedSources.size(), 1);
```

Reply:
@lawsonamzn So shouldn't only close out the file after we can be reasonably certain the AP has at least attempted the delete operation?

Starting the thread before the `AssetDeletedFile` can still lead to a race condition.

Reply:
This will not function as expected, BlockUntilIdle will either wait for the full time and return FALSE, or return early and become idle.

  The second block until idle will deadlock for all 5000 becuase at that point will already be idle with no further work to do.